### PR TITLE
[lexical-playground] Feature: Find and Replace

### LINKED ISSUE #8776: Feature: Find and Replace
## Description

One of the most powerful features in rich-text editors like Word and Google Docs is the **Find and Replace** function
Image

To search for text matches, you can often get by using the browser's built-in search function via <kbd>Ctrl</kbd> + <kbd>F</kbd>, but if you need to find and replace something you have to correct it manually or copy it into another editor and perform a global replace there

I think the main challenge with this functionality is finding text that lies at the intersection of several TextNodes and then extracting it. For example, if you're looking for the word “foo” to replace it with “bar”
```
 root
  └ paragraph 
    ├  text "f"
    └  text "oo" { format: bold }
```
When different formats are mixed the word should be replaced with a plain TextNode
```
 root
  └ paragraph 
    └  text "bar"
```

If the substring being searched for is contained within a single TextNode, the replacement must occur within that node, preserving its type, subclass, and format

## Impact

This will be helpful for any users who work with text documents

### PR DESCRIPTION
**Pull request by mayrang**
## Description

Adds a find-and-replace panel to the playground. The panel opens via Ctrl+F (find) or Cmd+Option+F (find and replace), and supports case-sensitive search, regex mode, match navigation, single replace, and replace-all.

Search runs against `registerTextContentListener`'s cached text — no tree traversal per keystroke, following the direction suggested in #8776. When matches exist, `$buildOffsetMap` walks the tree once via `$dfsWithSlotsIterator` to map global text offsets back to individual TextNode keys and local offsets. Highlights render through the CSS Custom Highlight API (`::highlight()`) on modern browsers, with a positioned-overlay-span fallback for browsers that don't support it. The dual-path approach mirrors `@lexical/yjs` SyncCursors.ts.

Replace uses `select()` + `setTextNodeRange()` + `insertText()` so that single-node replacements preserve the original text format (bold, italic, etc.). Replace-all processes matches back-to-front in a single `editor.update()` so earlier offsets stay valid.

### What changed (review feedback)

**1. Logic moved from React to Extension (`namedSignals` / commands)**

All search state (`searchTerm`, `replaceTerm`, `caseSensitive`, `isRegex`, `currentIndex`, `isOpen`) lives in `namedSignals` inside `FindReplaceExtension.build()`. Derived values (`matches`, `cachedText`) use `watchedSignal` and `computed`. Side effects (highlight rendering, match navigation, replace) run in `effect()` inside `register()`. The React panel is now a thin view — `useExtensionSignalValue` for reads, `editor.dispatchCommand()` for writes. The only remaining React hook is `useRef` for the search input focus ref.

**2. Regex capture group replacement (`$1`, `$2`, `$&`)**

`expandReplacement(template, matchText, searchRegex)` delegates to `String.prototype.replace`, so `$1`/`$2`/`$&`/`$$` all work natively. `findMatches` now returns `matchText` per match. Non-regex mode treats `$1` as a literal.

**3. Panel keyboard shortcuts**

Shortcuts that previously only worked with editor focus now also work when the panel itself has focus, via `onKeyDown` on the panel container:

| Shortcut | Action |
|---|---|
| Cmd+G / Ctrl+G | Next match |
| Shift+Cmd+G / Shift+Ctrl+G | Previous match |
| Cmd+F / Ctrl+F | Focus search input (blocks browser find) |
| Cmd+Option+F | Toggle panel (blocks browser default) |
| Enter / Shift+Enter | Next / previous match |
| Escape | Close panel |

**4. Scroll into view**

When navigating to a match that's off-screen, the match element scrolls into view (`scrollIntoView({behavior: 'smooth', block: 'nearest'})`). Works in both the CSS Highlights path and the overlay-span fallback.

### Design notes

- The offset map assumes `getTextContent()` inserts `\n\n` between sibling block elements and `\n` for `LineBreakNode`. If this convention changes in core, the map would need updating.
- Highlight names include `editor.getKey()` to avoid `CSS.highlights` collisions when multiple editors share a document.
- Invalid regex patterns show "Invalid regex" in the match counter rather than a silent "No results".

Closes #8776

## Test plan

- [x] `pnpm vitest run --project unit -- FindReplace` — 46 tests pass:
  - 14 `findMatches` + 3 zero-length regex edge cases (`.*`, `a*`, `x?`)
  - 7 `$buildOffsetMap`
  - 3 `$resolveMatchToPoints`
  - 6 `expandReplacement` (`$1`/`$2`, `$&`, `$$`, non-regex literal, no-group fallback)
  - 5 `$replaceMatch` (basic, format preservation, empty string, regex capture group, non-regex `$1` literal)
  - 4 `$replaceAllMatches` (basic, shorter, longer, regex capture group)
  - 4 Extension command dispatch integration (toggle, close, matches recompute, next/prev cycle)
- [x] e2e regression: `Ctrl+H` binding removed — `deleteBackward` tests pass (previously intercepted by the `Ctrl+H` handler)
- [x] tsc / eslint / prettier clean
- [x] Manual playground (Chrome, Firefox, Safari):
  - S1. Open/close: Cmd+F opens panel with search focused, Cmd+Option+F opens panel, Escape closes + clears highlights + restores editor focus, X button same as Escape
  - S2. Search + highlight: single match "1 / 1", multiple matches "1 / 3" with first in orange, no match "No results", live count update on text edit
  - S3. Next/previous navigation: Enter advances (1/3 → 2/3), Shift+Enter goes back, Cmd+G / Shift+Cmd+G from panel focus, Cmd+G from editor focus, wraps at end (3/3 → 1/3)
  - S4. Panel focus shortcuts: Cmd+F re-focuses search input (blocks browser find), Cmd+Option+F re-focuses (blocks browser default)
  - S5. Case sensitive / regex toggles: Aa toggle "Hello HELLO hello" → 3 → 1, regex toggle `\d+` → "No results" → "1 / 2", invalid regex `[invalid` → "Invalid regex"
  - S6. Single replace: "hello" → "goodbye" preserves surrounding text, empty replacement deletes matched text
  - S7. Replace all: 3 × "foo" → "qux" in one operation
  - S8. Regex capture groups: `(\w+)@(\w+)` + `$2/$1` → "user@host" → "host/user", `$&` whole-match reference, capture group replace-all, non-regex `$1` stays literal
  - S9. Formatted text: bold+plain boundary search highlights across nodes, replace inside bold preserves bold format
  - S10. Complex structures: multi-paragraph search, list item search
  - S11. Scroll into view: navigating to off-screen match scrolls it into the viewport

### CODE REVIEW

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 1]
```diff
@@ -0,0 +1,745 @@
...
+// Extension definitions
+// ---------------------------------------------------------------------------
+
+export const OPEN_FIND_REPLACE_COMMAND: LexicalCommand<void> =
+  /* @__PURE__ */ createCommand('OPEN_FIND_REPLACE_COMMAND');
+
+export const FindReplaceExtension = /* @__PURE__ */ defineExtension({
+  name: '@lexical/playground/FindReplace',
+  register: editor => {
+    return editor.registerCommand(
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: This works while the editor is selected but when in the find & replace dialog itself it defers to the native implementation. Might need to have a parallel set attached to the panel.

Probably also makes sense to register the other common shortcuts related to find & replace, particularly Cmd+G and Shift+Cmd+G for find next/previous but maybe some of the others? Screenshot from macOS Chrome

Image

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+      COMMAND_PRIORITY_LOW,
+    );
+  },
+});
+
+// ---------------------------------------------------------------------------
+// React UI
+// ---------------------------------------------------------------------------
+
+function FindReplacePanel({
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: It would be a better abstraction if this was really just the UI (none of the useEffect boilerplate here) and the logic was in the extension, maybe facilitated by commands. Would be easier to port.

**Review by etrepum — COMMENTED** [review]
This is very cool! For regex find+replace I would expect $1, $2, etc. in the replacement to reference the groups in the source

**Comment by mayrang** [comment]
Done — `expandReplacement` delegates to `String.prototype.replace` so `$1`/`$2`/`$&`/`$$` all work natively. Non-regex mode treats `$1` as a literal.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 1 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Done — panel `onKeyDown` handles Cmd+G / Shift+Cmd+G, Cmd+F, Ctrl+H (non-macOS) / Cmd+Option+F, Enter / Shift+Enter, and Escape. `Ctrl+H` skipped on macOS (Emacs `deleteBackward`).

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 2 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Done — all state in `namedSignals`, derived values via `watchedSignal`/`computed`, side effects in `effect()`. Panel is `useExtensionSignalValue` + `dispatchCommand`, only remaining hook is `useRef`.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+  format: number;
+}
+
+// ---------------------------------------------------------------------------
+// Pure functions
+// ---------------------------------------------------------------------------
+
+const EMPTY_MATCHES: TextMatch[] = Object.freeze<TextMatch[]>(
+  [],
+) as TextMatch[];
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Probably better in this case to just `return []` from findMatches rather than to lie and say that this readonly value is actually mutable.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 4]
```diff
@@ -0,0 +1,987 @@
...
+      caseSensitive: false,
+      currentIndex: 0,
+      isOpen: false,
+      isRegex: false,
+      replaceTerm: '',
+      searchTerm: '',
+    });
+
+    const cachedText = watchedSignal(
+      () => editor.read(() => $getRoot().getTextContent()),
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: I don't think we want this to force a commit

```suggestion
      () => editor.read('latest', () => $getRoot().getTextContent()),
```

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 5]
```diff
@@ -0,0 +1,987 @@
...
+          output.currentIndex.value = 0;
+          return true;
+        },
+        COMMAND_PRIORITY_LOW,
+      ),
+
+      editor.registerCommand(
+        KEY_DOWN_COMMAND,
+        event => {
+          const isFindReplace =
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: These would probably be a little better with isExactShortcutMatch and/or isModifierMatch

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx [hunk 6]
```diff
@@ -0,0 +1,987 @@
...
+    }
+    if (e.key === 'Enter') {
+      e.preventDefault();
+      editor.dispatchCommand(
+        e.shiftKey ? FIND_PREV_COMMAND : FIND_NEXT_COMMAND,
+        undefined,
+      );
+      return;
+    }
+    const isMod = IS_APPLE ? e.metaKey && !e.ctrlKey : e.ctrlKey && !e.metaKey;
```
Comment (etrepum) [inline packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Looks like more code that could benefit from isExactShortcutMatch

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 3 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Done — returning [] directly.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 4 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Done — switched to editor.read('latest', ...).

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 5 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Done — extracted a CONTROL_OR_META constant and switched both KEY_DOWN_COMMAND and panel onKeyDown to isExactShortcutMatch.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 6 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Covered in the same change as above.

File: packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx (hunk 1 above)
Comment (mayrang) [reply packages/lexical-playground/src/plugins/FindReplaceExtension/index.tsx]: Correction: removed `Ctrl+H` binding entirely. macOS maps `Ctrl+H` to Emacs `deleteBackward`, and Playwright sends it as the `deleteBackward` shortcut — the binding broke 9 e2e tests. Only Cmd+F (find) and Cmd+Option+F (find+replace) open the panel now.

**Review by etrepum — COMMENTED** [review]
Overall this works great, here's two small improvements I think would be nice before merge:

The replace algorithm should pick the smallest replacement target in terms of document coordinates, e.g. if you open https://lexical-playground-git-fork-mayrang-feat-87-a6242b-fbopensource.vercel.app/?emptyEditor=false and replace "repository" with "repo" it lifts the TextNode out of the LinkNode (same behavior for replacing at the leading edge of a LinkNode). I would expect it to replace just the text so "repo" is still in the LinkNode.

The tab index could match a more typical find & replace dialog's order, most importantly I expect pressing tab from "find" to take me to "replace" before visiting the other controls. vscode screen recording below

https://github.com/user-attachments/assets/fd724eed-2b58-4738-87c7-0d9b7a69ffd4

**Comment by mayrang** [comment]
Both fixed. Single-node matches now use `spliceText` so the replacement stays inside its parent (LinkNode etc.) instead of being lifted out by `selection.insertText`. Panel layout switched to CSS Grid — DOM order controls tab sequence (Find → Replace → Aa → .* → Prev → Next → Replace → All → Close) while grid placement keeps the visual arrangement. Added a Tab key interceptor for Safari since it skips buttons by default.

**Review by etrepum — APPROVED** [review]
